### ISSUE TITLE
Further questions on the attention kernels

### ISSUE THREAD

**Issue by rayleizhu** [NVIDIA/TensorRT-LLM#584]
1. Does the [generation-phase kernel (decoderMaskedMultiheadAttention)](https://github.com/NVIDIA/TensorRT-LLM/tree/release/0.5.0/cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention) support query length > 1 (e.g., the query has a logical shape of (bs=8, head=8, q_len=32, hed_dim=64))?   
    - As far as I know, for the GPT inference scenario, q_len is always 1 in the generation phase.
    - However, my hacking needs to cope with q_len > 1. Is the generation-phase kernel general, like FlashDecoding, or specialized for q_len=1?
    - If the generation-phase kernel does not support q_len > 1, can you give some suggestions for my hacking?

2. It seems that the [context-phase kernel (contextFusedMultiHeadAttention)](https://github.com/NVIDIA/TensorRT-LLM/tree/release/0.5.0/cpp/tensorrt_llm/kernels/contextFusedMultiHeadAttention) is provided with binary code. This makes it not hackable. Could you let me know if you're going to release the source code in the near future?

Related issue: https://github.com/NVIDIA/TensorRT-LLM/issues/457

**Comment by byshiue** [comment]
@ncomly-nvidia  Could you help to answer the question 2?

**Comment by PerkzZheng** [comment]
@rayleizhu answer you questions,

1.  it doesn't support query length > 1 explicitly. As a workaround, you can take this case as GQA (group query attention) where multiple q heads (i.e. query length in your case) share the same kv heads. And that is supported by the MMHA kernel.  At the same time, we have another optimized kernel for GQA, and you can take a look at [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/plugins/gptAttentionCommon/gptAttentionCommon.cpp#L1091) to see if your case is supported. If not, you can fall back to the MMHA kernel.

3. there is no plan to release the source code currently.

**Comment by YavorGIvanov** [comment]
It seems that the multi-block mode part of the decoderMaskedMultiheadAttention kernel is actively worked on as there are changes in the previous updates 
- https://github.com/NVIDIA/TensorRT-LLM/pull/613/files#diff-a0fc455396e3396bcedc04686d26721ec0a84218c3c264c072aeb7f7f706e08e
- https://github.com/NVIDIA/TensorRT-LLM/pull/667/files#diff-a0fc455396e3396bcedc04686d26721ec0a84218c3c264c072aeb7f7f706e08e

Are its results more stable now ? What have you observed when testing with GPT or similar model ?

**Comment by PerkzZheng** [comment]
@YavorGIvanov 
in terms of stability of the multi-block mode, we have fixed several bugs related to that, so definitely try out the latest main branch if you meet accuracy issues due to that. We have seen no accuracy drop even with up to 32K sequence length (llama 70B) with the recent fix in the main branch. Let me know if you severe accuracy drop with multi-block mode.

**Comment by YavorGIvanov** [comment]
> @rayleizhu answer you questions,
> 
> 1. it doesn't support query length > 1 explicitly. As a workaround, you can take this case as GQA (group query attention) where multiple q heads (i.e. query length in your case) share the same kv heads. And that is supported by the MMHA kernel.  At the same time, we have another optimized kernel for GQA, and you can take a look at [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/plugins/gptAttentionCommon/gptAttentionCommon.cpp#L1091) to see if your case is supported. If not, you can fall back to the MMHA kernel.
> 2. there is no plan to release the source code currently.

For 1.

if I have say 3 input prompts of 100 tokens and 32 heads with 64 dim per head (2048 hidden dimension) then what would be the appropriate values for these parameters if I was to use GQA (Group Querry Attention) by the MMHA kernel for:
- stride
- batch_size
- num_heads
- num_kv_heads

The QKV buffer is linear and is [300, 6144]
I tried using stride 0, batch_size 3, num_heads 3200 = 32 * 100, num_kv_heads 32, sequence_length 1 (Can't be 0 as you subtract 1 in the kernel)

With stride 0 the q_stride in the kernel is calculated by 32 * 64 = 2048, however as the buffer is linear shouldn't it actually
be 2048 * 3 ? 

Trying the above parameters results in incorrect memory accesses and even changing the q_stride to be harcoded to 2048 * 3 doesn't help. 
I am misunderstanding something and I can't find an example of GQA with the MMHA kernel as the XQA is used in the gptAttentionPlugin.

**Comment by PerkzZheng** [comment]
after rethinking, directly using MMHA kernels is not goitng  to work as we assume key/value length is 1 for the new tokens. 
you have to add another loop in this [part](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h#L1460-L1739) to read more than one key/value tokens.

XQA won't work for the same reason. We are working on supporting that feature with XQA.

**Comment by YavorGIvanov** [comment]
I have one additional question about the MMHA kernel. 
- Does it support decoding a batch with input prompts, which have different lengths ? I know it isn't used for query length > 1, but I am trying to process the input prompts (context step) with different kernel and then use the MMHA kernel as part of decoding the next single tokens. 
- I see that the "input_lengths" parameter is only used for beam searching and the "lengths_per_sample" is expected to be "max_input_length + current output length", so if we have two input prompts (e.g. one with length 20 and another with length 100)  and I provide "lengths_per_samples" containing [101, 101], how will it know that in the KV cache for one the batches there is only 20 token entries and for the other there is 100 ?

**Comment by PerkzZheng** [comment]
> * Does it support decoding a batch with input prompts, which have different lengths ? I know it isn't used for query length > 1, but I am trying to process the input prompts (context step) with different kernel and then use the MMHA kernel as part of decoding the next single tokens.

Yes, it has been supported since the first release of TRT-LLM, and it is implemented based on the `length_per_sample`.

> * I see that the "input_lengths" parameter is only used for beam searching and the "lengths_per_sample" is expected to be "max_input_length + current output length", so if we have two input prompts (e.g. one with length 20 and another with length 100)  and I provide "lengths_per_samples" containing [101, 101], how will it know that in the KV cache for one the batches there is only 20 token entries and for the other there is 100 ?

`lengths_per_sample = actual_input_length + current_output_length`.  `actual_input_length` can be variable across sequences in one batch while `current_output_length` should be the same for all sequences. So `lengths_per_samples` should be `[21, 101]` instead, and 21 means there are (21 - 1 = 20) tokens in the past kv cache, and we will add one new token into the kv cache.

**Comment by YavorGIvanov** [comment]
That was very useful. Thank you for your work and help. 🙏🏻 Once I get to testing and benchmarking multi-block mode, I will provide detailed feedback, so I can give back :)   

This is the comment that misled me https://github.com/NVIDIA/TensorRT-LLM/blob/d879430b040ea30c65220c9171a8a70607398748/cpp/tensorrt_llm/plugins/gptAttentionCommon/gptAttentionCommon.cpp#L255

**Comment by PerkzZheng** [comment]
> This is the comment that misled me

Thanks. I will modify that comment.
